src/plugin-host/plugin/plugin.ts

```diff
@@ -0,0 +1,139 @@
+{
+  /**
+   * Semantic identifier of this plugin. 
+   */
+  public Name: string;
+  /**
+   * Version of this plugin.
+   */
+  public Version: string;
+  /**
+   * Boolean indicating the status of this plugin.
+   * Disabled plugins will not receive messages
+   * from its Plugin Host.
+   */
+  public Enabled: boolean;
+  /**
+   * Internal plugin state.
+   */
+  protected Data: any;
+
+  /**
+   * Services for this chat
+   */
+  public Services: () => ChatServices;
+
+  /**
+   * Plugin ID
+   */
+  public PID: () => string;
+
+  /**
+   * Event triggers. Plugins can hook functions to certain Plugin Events.
+   * These plugin events are defined in the PLUGIN_EVENT enumeration.
+   */
+  private pluginEventTriggers: Map<PLUGIN_EVENT, (data: any) => any>;
+
+  /**
+   * Command triggers. Plugins can hook functions to certain commands.
+   * these commands can be set in the constructor of a plugin.
+   */
+  private pluginCommandTriggers: PluginCommand[];
+
+  /**
+   * Create a new Plugin instance.
+   * @param _name Semantic name of this plugin.
+   * @param _version Version of this plugin.
+   * @param _data Any optional data that might require to be stored
+   *              for this plugin.
+   */
+  constructor(_name: string, _version: string, _data: any) 
+  {
+    this.Name = _name;
+    this.Version = _version;
+    this.Data = _data;
+    this.pluginEventTriggers = new Map<PLUGIN_EVENT, (data: any) => any>();
+    this.pluginCommandTriggers = [];
+
+    console.log("+ Loaded plugin: " + this.Name + " (" + this.PID + ")");
+  };
+
+  /* Function overload list */
+  protected subscribeToPluginEvent(_event: PLUGIN_EVENT.PLUGIN_EVENT_PRE_MESSAGE, _eventFn: (_data: PrePostMessagePluginEventArguments) => any): void;
+  protected subscribeToPluginEvent(_event: PLUGIN_EVENT.PLUGIN_EVENT_POST_MESSAGE, _eventFn: (_data: PrePostMessagePluginEventArguments) => any): void;
+  protected subscribeToPluginEvent(_event: PLUGIN_EVENT.PLUGIN_EVENT_USER_CHANGED_SCORE, _eventFn: (_data: UserScoreChangedPluginEventArguments) => any): void;
+  protected subscribeToPluginEvent(_event: PLUGIN_EVENT.PLUGIN_EVENT_LEADERBOARD_RESET, _eventFn: (_data: LeaderboardResetPluginEventArguments) => any): void;
+  protected subscribeToPluginEvent(_event: PLUGIN_EVENT.PLUGIN_EVENT_DANKTIMES_SHUTDOWN, _eventFn: (_data: NoArgumentsPluginEventArguments) => any): void;
+  protected subscribeToPluginEvent(_event: PLUGIN_EVENT.PLUGIN_EVENT_POST_INIT, _eventFn: (_data: NoArgumentsPluginEventArguments) => any): void;
+  /**
+   * Subscribe to a certain PLUGIN_EVENT.
+   * @param _event Plugin event to describe to.
+   * @param _eventFn Function to execute when a certain event is triggered.
+   */
+  protected subscribeToPluginEvent(_event: PLUGIN_EVENT, _eventFn: (_data: any) => any): void
+  {
+    this.pluginEventTriggers.set(_event, _eventFn);
+  }
+
+  /**
+   * Register a new plugin commands.
+   * These are custom /commands that can be invoked
```

Comment:
How does this handle duplicate command names?

Reply:
Duplicates can be registered, but the command will only be run once. This should be the first one that was registered.

```ts
    let trigger = this.pluginCommandTriggers.find(commands => commands.commandString === command);
```